### DISCUSSION TITLE
Improving Functional coverage performance

### DISCUSSION THREAD

**Discussion by dobios** [chiselverify/chiselverify#39] [Ideas]
A main problem with the current state of the functional coverage tool is that it stores all of its data inside a large amount of `HashMap`s.   
The problem with this is that mutable HashMaps in Scala are inherently slow, more importantly, they are a generic solution made for quickly solving a lot of problems.  
  
An analysis of the performance of HashMaps can be found in this [very useful stackoverflow post](https://stackoverflow.com/questions/28744990/why-is-scala-hashmap-slow/28747113).      

__TLDR of the post__: HashMaps are slow and different collections should be used based on the type of the key being used.    
  
### Applying this to ChiselVerify  
In our case, most of our keys can be made into either `Int` keys or `String` keys. This means that the HashMaps should be changed to:  
  1. __String Keys__: `AnyRefMap` 
  2. __Int Keys__: `LongMap`
